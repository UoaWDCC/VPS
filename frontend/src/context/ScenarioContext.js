import React from "react";

const ScenarioContext = React.createContext();

export default ScenarioContext;
