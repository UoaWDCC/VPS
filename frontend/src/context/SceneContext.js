import React from "react";

const SceneContext = React.createContext();

export default SceneContext;
