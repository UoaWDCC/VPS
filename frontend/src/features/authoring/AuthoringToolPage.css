body {
  background-color: #222;
}

.actions {
  display: flex;
  gap: 6px;
  padding: 5px;
  color: white;
  align-items: center;
  margin-left: 20px;
}

.topbar {
  background-color: #151515;
  width: 100%;
  z-index: 1;
}

.props {
  display: flex;
  padding: 5px;
  flex-wrap: wrap;
  gap: 6px;
  color: white;
  align-items: center;
  margin-left: 20px;
}

.props > div {
  display: flex;
  color: white;
  align-items: center;
}

.button {
  position: relative;
  width: 28px;
  height: 28px;
  border-radius: 5px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.button:hover {
  background-color: #0f0f0f;
}

.text-input {
  background-color: #000;
  border-radius: 5px;
  padding: 0 10px;
  height: 28px;
}

.color-input {
  position: relative;
  width: 18px;
  height: 18px;
  border-bottom: 3px solid white;
  display: flex;
  justify-content: center;
  align-items: center;
}

.color-input span {
  font-size: 14px;
}

.button.active {
  background-color: #000;
}

input[type="number"]::-webkit-outer-spin-button,
input[type="number"]::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

input[type="number"] {
  -moz-appearance: textfield;
}
