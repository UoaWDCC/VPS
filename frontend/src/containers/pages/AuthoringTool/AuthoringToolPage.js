import React, { useContext, useEffect, useState } from "react";
import { useHistory, useParams } from "react-router-dom";
import { Button } from "@material-ui/core";
import TopBar from "../../../components/TopBar";
import ToolBar from "./ToolBar/ToolBar";
import Canvas from "./Canvas/Canvas";
import CanvasSideBar from "./CanvasSideBar/CanvasSideBar";
import ScreenContainer from "../../../components/ScreenContainer";
import ScenarioContext from "../../../context/ScenarioContext";
import { useGet, usePut } from "../../../hooks/crudHooks";
import SceneContext from "../../../context/SceneContext";
import AuthoringToolContext from "../../../context/AuthoringToolContext";
import ToolbarContextProvider from "../../../context/ToolbarContextProvider";
import AuthenticationContext from "../../../context/AuthenticationContext";
import { uploadFiles } from "../../../firebase/storage";
import HelpButton from "../../../components/HelpButton";
import SceneNavigator from "./SceneNavigator/SceneNavigator";

/**
 * This page allows the user to edit a scene.
 * @container
 */
export default function AuthoringToolPage() {
  const { scenarioId, sceneId } = useParams();
  const {
    scenes,
    currentScene,
    setCurrentScene,
    setMonitorChange,
    setHasChange,
    reFetch,
  } = useContext(SceneContext);
  const { currentScenario } = useContext(ScenarioContext);
  const { setSelect } = useContext(AuthoringToolContext);
  const { getUserIdToken } = useContext(AuthenticationContext);
  const [firstTimeRender, setFirstTimeRender] = useState(true);
  const [saveButtonText, setSaveButtonText] = useState("Save");
  const [disableNext, setDisableNext] = useState(false);
  const [disableBefore, setDisableBefore] = useState(false);
  const history = useHistory();
  useGet(
    `/api/scenario/${currentScenario?._id}/scene/full/${currentScene?._id}`,
    setCurrentScene,
    false
  );

  useEffect(async () => {
    reFetch();
  }, []);

  useEffect(() => {
    if (firstTimeRender) {
      setFirstTimeRender(false);
    } else {
      setMonitorChange(true);
    }
  }, [currentScene]);

  /** used to save the scene, as a helper function */
  async function saveScene() {
    setSelect(null);
    await uploadFiles(
      currentScene?.components,
      currentScenario._id,
      currentScene._id
    );
    await usePut(
      `/api/scenario/${scenarioId}/scene/${sceneId}`,
      {
        name: currentScene.name,
        components: currentScene?.components,
      },
      getUserIdToken
    );
    setHasChange(false);
    reFetch();
  }

  /** called when save button is clicked */
  async function save() {
    saveScene();
    setSaveButtonText("Saved!");
    setTimeout(() => {
      setSaveButtonText("Save");
    }, 1800);
  }

  /** called when save and close button is clicked */
  function savePlusClose() {
    saveScene();
    /* redirects user to the scenario page */
    window.location.href = `/scenario/${currentScenario?._id}`;
  }

  function updateScene(direction) {
    const ids = scenes.map((scene) => scene._id);
    const oldIndex = ids.indexOf(currentScene._id);
    const newIndex = oldIndex + direction;

    if (newIndex > -1 && newIndex < ids.length) {
      saveScene();
      setDisableBefore(false);
      setDisableNext(false);
      setCurrentScene(scenes[newIndex]);
      history.push({
        pathname: `/scenario/${scenarioId}/scene/${scenes[newIndex]._id}`,
      });
    }

    if (newIndex === ids.length - 1) {
      setDisableNext(true);
    }

    if (newIndex === 0) {
      setDisableBefore(true);
    }
  }

  return (
    <>
      <ScreenContainer vertical>
        <TopBar back={`/scenario/${currentScenario?._id}`} confirmModal>
          <Button
            className="btn top contained white"
            color="default"
            variant="contained"
            onClick={save}
          >
            {saveButtonText}
          </Button>
          <Button
            className="btn top contained white"
            color="default"
            variant="contained"
            onClick={savePlusClose}
          >
            Save & close
          </Button>
          <Button
            className={`btn top contained white ${
              disableBefore ? "disabled" : ""
            }`}
            color="default"
            variant="contained"
            disabled={disableBefore}
            onClick={() => updateScene(-1)}
          >
            Before
          </Button>
          <Button
            className={`btn top contained white ${
              disableNext ? "disabled" : ""
            }`}
            color="default"
            variant="contained"
            disabled={disableNext}
            onClick={() => updateScene(1)}
          >
            Next
          </Button>
          <HelpButton />
        </TopBar>
        <ToolbarContextProvider>
          <ToolBar />
        </ToolbarContextProvider>
        <div className="flex" style={{ height: "100%", overflow: "hidden" }}>
          <SceneNavigator saveScene={saveScene} />
          <Canvas />
          <CanvasSideBar />
        </div>
      </ScreenContainer>
    </>
  );
}
