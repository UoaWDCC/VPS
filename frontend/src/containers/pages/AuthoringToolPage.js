import React, { useContext } from "react";
import TopBar from "../../components/TopBar";
import ScreenContainer from "../../components/ScreenContainer";
import SceneContext from "../../context/SceneContext";
import ScenarioContext from "../../context/ScenarioContext";
import { useGet } from "../../hooks/crudHooks";

export default function AuthoringToolPage() {
  const { currentScene, setCurrentScene } = useContext(SceneContext);
  const { currentScenario } = useContext(ScenarioContext);
  useGet(
    `/api/scenario/${currentScenario?._id}/scene/full/${currentScene?._id}`,
    setCurrentScene
  );

  return (
    <>
      <ScreenContainer vertical>
        <TopBar back={`/scenario/${currentScenario?._id}`} />
        <h1>Authoring Tool Page</h1>
      </ScreenContainer>
    </>
  );
}
