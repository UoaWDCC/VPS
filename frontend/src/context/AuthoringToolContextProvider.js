import React, { useState, useContext } from "react";
import AuthoringToolContext from "./AuthoringToolContext";
import SceneContext from "./SceneContext";

export default function AuthoringToolContextProvider({ children }) {
  const { currentScene, setCurrentScene } = useContext(SceneContext);

  const [select, setSelect] = useState(null);
  const [bounds, setBounds] = useState(null);
  const [shiftPressed, setShiftPressed] = useState(false);

  function selectElement({ currentTarget }) {
    const image = currentTarget.firstElementChild?.nodeName === "IMG";
    setSelect(image ? currentTarget.firstElementChild.id : currentTarget.id);
  }

  const clearElement = ({ target }) => {
    if (target.id === "canvas") {
      setSelect(null);
    }
  };

  function deleteElement() {
    if (select !== null) {
      console.log(currentScene.components);
      console.log(select);
      const updatedComponents = currentScene.components;
      console.log(updatedComponents.splice(select, 1));

      console.log(updatedComponents);

      setSelect(null);

      setCurrentScene({
        ...currentScene,
        components: updatedComponents,
      });
    }
  }

  return (
    <AuthoringToolContext.Provider
      value={{
        select,
        setSelect,
        selectElement,
        clearElement,
        bounds,
        setBounds,
        shiftPressed,
        setShiftPressed,
        deleteElement,
      }}
    >
      {children}
    </AuthoringToolContext.Provider>
  );
}
