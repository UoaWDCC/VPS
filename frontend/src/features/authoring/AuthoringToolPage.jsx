import { useContext, useEffect, useState, useRef } from "react";
import { useParams } from "react-router-dom";
import HelpButton from "components/HelpButton";
import ScreenContainer from "components/ScreenContainer/ScreenContainer";
import TopBar from "components/TopBar/TopBar";
import AuthenticationContext from "context/AuthenticationContext";
import AuthoringToolContext from "context/AuthoringToolContext";
import ScenarioContext from "context/ScenarioContext";
import SceneContext from "context/SceneContext";
import ToolbarContextProvider from "context/ToolbarContextProvider";
import { uploadFiles } from "../../firebase/storage";
import { useGet, usePut } from "hooks/crudHooks";
import Canvas from "./Canvas/Canvas";
import CanvasSideBar from "./CanvasSideBar/CanvasSideBar";
import SceneNavigator from "./SceneNavigator/SceneNavigator";
import ToolBar from "./ToolBar/ToolBar";

/**
 * This page allows the user to edit a scene.
 * @container
 */
export default function AuthoringToolPage() {
  const { scenarioId, sceneId } = useParams();
  const {
    currentScene,
    setCurrentScene,
    setMonitorChange,
    setHasChange,
    reFetch,
  } = useContext(SceneContext);
  const { currentScenario } = useContext(ScenarioContext);
  const { setSelect } = useContext(AuthoringToolContext);
  const { getUserIdToken } = useContext(AuthenticationContext);
  const [firstTimeRender, setFirstTimeRender] = useState(true);
  const [saveButtonText, setSaveButtonText] = useState("Save");
  const autosaveTimeout = useRef(null);

  useGet(
    `/api/scenario/${currentScenario?._id}/scene/full/${currentScene?._id}`,
    setCurrentScene,
    true
  );

  useEffect(async () => {
    reFetch();
  }, []);

  useEffect(() => {
    if (firstTimeRender) {
      setFirstTimeRender(false);
    } else {
      setMonitorChange(true);
    }
  }, [currentScene]);

  useEffect(() => {
    if (!currentScene || !currentScene._id) return;
    if (firstTimeRender) return;
    if (autosaveTimeout.current) clearTimeout(autosaveTimeout.current);
    autosaveTimeout.current = setTimeout(() => {
      saveScene(true);
    }, 2000);
    return () => clearTimeout(autosaveTimeout.current);
  }, [currentScene]);

  /** used to save the scene, as a helper function */
  async function saveScene(isAuto = false) {
    if (!isAuto) {
      // only clear selection on manual save
      setSelect(null);
    }
    await uploadFiles(
      currentScene?.components,
      currentScene?.endImage,
      currentScenario._id,
      currentScene._id
    );
    await usePut(
      `/api/scenario/${scenarioId}/scene/${sceneId}`,
      {
        name: currentScene.name,
        time: currentScene.time,
        components: currentScene?.components,
      },
      getUserIdToken
    );
    setHasChange(false);
    reFetch();
  }

  /** called when save button is clicked */
  async function save() {
    await saveScene();
    setSaveButtonText("Saved!");
    setTimeout(() => {
      setSaveButtonText("Save");
    }, 1800);
  }

  /** called when save and close button is clicked */
  async function savePlusClose() {
    await save();
    /* redirects user to the scenario page */
    window.location.href = `/scenario/${currentScenario?._id}`;
  }

  return (
    <>
      <ScreenContainer vertical>
        <TopBar back={`/scenario/${currentScenario?._id}`} confirmModal>
          <button className="btn vps w-[150px]" onClick={save}>
            {saveButtonText}
          </button>
          <button className="btn vps w-[150px]" onClick={savePlusClose}>
            Save & Close
          </button>
          <HelpButton />
        </TopBar>
        <ToolbarContextProvider>
          <ToolBar />
        </ToolbarContextProvider>
        <div className="flex" style={{ height: "100%", overflow: "hidden" }}>
          <SceneNavigator saveScene={saveScene} />
          <Canvas />
          <CanvasSideBar />
        </div>
      </ScreenContainer>
    </>
  );
}
