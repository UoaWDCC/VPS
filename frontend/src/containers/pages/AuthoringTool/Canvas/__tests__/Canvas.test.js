import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import AuthoringToolPage from "../../AuthoringToolPage";
import ScenarioContext from "../../../../../context/ScenarioContext";
import SceneContext from "../../../../../context/SceneContext";
import AuthoringToolContextProvider from "../../../../../context/AuthoringToolContextProvider";

test("Authoring Tool page snapshot test", () => {
  const context = {
    currentScenario: { _id: "scenarioId" },
    currentScene: { _id: "sceneId", components: [] },
    setMonitorChange: () => {},
  };

  const { baseElement } = render(
    <BrowserRouter>
      <ScenarioContext.Provider value={context}>
        <SceneContext.Provider value={context}>
          <AuthoringToolContextProvider>
            <AuthoringToolPage />
          </AuthoringToolContextProvider>
        </SceneContext.Provider>
      </ScenarioContext.Provider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});
