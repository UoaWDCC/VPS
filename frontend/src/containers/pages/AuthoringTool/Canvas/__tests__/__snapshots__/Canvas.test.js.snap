// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authoring Tool page snapshot test 1`] = `
<body>
  <div>
    <div
      class="moveable-control-box    rCSuehiy8"
      data-able-draggable="true"
      data-able-origin="true"
      data-able-padding="true"
      data-able-resizable="true"
      data-able-snappable="true"
      data-styled-id="rCSuehiy8"
      style="position: absolute; display: none; transform: translate3d(0px, 0px, 0); --zoom: 1; --zoompx: 1px;"
    >
      <div
        class="moveable-control moveable-direction moveable-n "
        data-direction="n"
        data-rotation="0"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-w "
        data-direction="w"
        data-rotation="90"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-s "
        data-direction="s"
        data-rotation="0"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-e "
        data-direction="e"
        data-rotation="90"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-nw "
        data-direction="nw"
        data-rotation="135"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-ne "
        data-direction="ne"
        data-rotation="45"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-sw "
        data-direction="sw"
        data-rotation="45"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-direction moveable-se "
        data-direction="se"
        data-rotation="135"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-control moveable-origin"
        style="transform: translateZ(0px) translate(0px, 0px) rotate(0rad) scale(1);"
      />
      <div
        class="moveable-line moveable-direction "
        data-direction=""
        data-line-index="0"
        data-rotation="-1"
        style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
      />
      <div
        class="moveable-line moveable-direction "
        data-direction=""
        data-line-index="1"
        data-rotation="-1"
        style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
      />
      <div
        class="moveable-line moveable-direction "
        data-direction=""
        data-line-index="2"
        data-rotation="-1"
        style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
      />
      <div
        class="moveable-line moveable-direction "
        data-direction=""
        data-line-index="3"
        data-rotation="-1"
        style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"
      />
    </div>
    <div
      class="canvasContainer"
    >
      <div
        class="canvas"
        id="canvas"
      />
    </div>
  </div>
</body>
`;
